File: divide_conquer/01_dc_longest_common_prefix/test_dc_longest_common_prefix.cpp
#include "dc_longest_common_prefix.cpp"
#include <gtest/gtest.h>
#include <chrono>
#include <iostream>
#include <vector>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestDCLongestCommonPrefix : public ::testing::Test {
protected:
    Solution* solution; // Each test will have its own pointer to Solution instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        solution = new Solution();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete solution;
    }
};


TEST_F(TestDCLongestCommonPrefix, test1_fl) { 
    vector<string> strs = {"flower", "flow", "flight"};
    string expected = "fl";
    string result = solution->longestCommonPrefix(strs);

    ASSERT_EQ(expected, result);
}

TEST_F(TestDCLongestCommonPrefix, test2_empty) { 
    vector<string> strs = {"dog", "racecar", "car"};
    string expected = "";
    string result = solution->longestCommonPrefix(strs);

    ASSERT_EQ(expected, result);
}


int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}

File: divide_conquer/01_dc_longest_common_prefix/test_dc_longest_common_prefix.py
from unittest import TestCase, TestResult, TestLoader, TextTestRunner
from dc_longest_common_prefix import Solution
from time import time

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


class TestDCLongestCommonPrefix(TestCase):
    def setUp(self):
        # This method is called before each test
        self.solution = Solution()

    def test1_fl(self):
        strs = ["flower", "flow", "flight"]
        expected = "fl"
        result = self.solution.longest_common_prefix(strs)

        self.assertEqual(result, expected)

    def test2_dogcar(self):
        strs = ["dog", "racecar", "cat"]
        expected = ""
        result = self.solution.longest_common_prefix(strs)

        self.assertEqual(result, expected)

    def test3_flower(self):
        strs = ["flower", "flower", "flower", "flower"]
        expected = "flower"
        result = self.solution.longest_common_prefix(strs)

        self.assertEqual(result, expected)

    def test4_empty(self):
        strs = [""]
        expected = ""
        result = self.solution.longest_common_prefix(strs)

        self.assertEqual(result, expected)

    def test5_abc(self):
        strs = ["abca", "abc"]
        expected = "abc"
        result = self.solution.longest_common_prefix(strs)

        self.assertEqual(result, expected)

    def tearDown(self):
        del self.solution  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestDCLongestCommonPrefix)
    result = CustomTestResult()
    runner = TextTestRunner(verbosity=2)
    # runner.run(suite)
    suite.run(result)
    result.printTestTimes()
